dissect/target/helpers/regutil.py

```diff
@@ -183,6 +183,7 @@ class VirtualHive(RegistryHive):
 
     def __init__(self):
         self._root = VirtualKey(self, "")
+        self.filepath = None
```

Comment:
As you can see below, the function `self.target.registry.get_user_details(key)` breaks when using a `VirtualHive`. Adding `self.filepath = None` fixes this for virtual testing using pytest.

```
self = <dissect.target.plugins.os.windows.registry.RegistryPlugin object at 0x7f559a5ee690>, key = <VirtualKey PuTTY>

    @internal
    def get_user_details(self, key: RegistryKey) -> UserDetails:
        """Return user details for the user who owns a registry hive that contains the provided key"""
>       if not key.hive or not key.hive.filepath:
E       AttributeError: 'VirtualHive' object has no attribute 'filepath'

dissect/target/plugins/os/windows/registry.py:332: AttributeError

Reply:
This is a band-aid fix. This should either be fixed in the base `RegistryHive` class, or with a `hasattr` in that location.

Reply:
The `RegfHive` class sets `self.filepath` in a similar fashion during `__init__`. I assumed you would want the  `VirtualHive` class to look like a `RegfHive` class as closely as possible. The assumption that a `hive` has the `filepath` attribute is not inherently wrong in my opinion. It seems that dissect assumes the `VirtualHive` class can be used as a drop-in replacement with `RegfHive`, which is currently not true.

What is your opinion on the "api" structure of `RegfHive` implementations? Would you propose setting `self.filepath` to something else than `None` or still go for the `hasattr` check in the `RegistryPlugin`?

Reply:
You might be correct, but it's a band-aid solution to put that attribute in `VirtualHive`. It only gets rid of this current error, but it doesn't fix the underlying issue. Because what happens if I use a different subclass from `RegistryHive`, such as the one from the Carbon Black or SMB loader? Now I'm adding a attribute to that subclass to avoid the specific error for that specific class.

So the correct solution is to either make it an attribute on the base `RegistryHive` class, test for the specific type of `RegfHive` class with `isinstance`, or test for the availability of the attribute (duck-typing, commonly used in Python).

Personally I'm leaning towards going the duck-typing route, because as of right now there's little reason to add a generic `filepath` attribute to those other implementations.

Reply:
Thanks for the explanation. Added duck typing check in [9ebb0a1](https://github.com/fox-it/dissect.target/pull/418/commits/9ebb0a1eb4dd6c11da7160fbaa47029be10dc492).